processing/src/main/java/org/apache/druid/query/filter/FallbackPredicate.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.filter;
+
+import com.google.common.base.Predicate;
+import org.apache.druid.math.expr.ExprEval;
+import org.apache.druid.math.expr.ExpressionType;
+
+import javax.annotation.Nullable;
+
+/**
+ * {@link Predicate} that wraps another predicate and can catch {@link ClassCastException} from some
+ * {@link java.util.Comparator} and degrades into casting input to an expected {@link ExpressionType} once an exception
+ * is encountered. Useful when processing data that might be mixed types, despite what the column capabilities might
+ * claim the type is, such as variant 'auto' types. This class is not thread-safe.
+ */
+public class FallbackPredicate<T> implements Predicate<T>
+{
+  private final Predicate<T> delegate;
+  private final ExpressionType expectedType;
+  private boolean needsCast = false;
+
+  public FallbackPredicate(Predicate<T> delegate, ExpressionType expectedType)
+  {
+    this.delegate = delegate;
+    this.expectedType = expectedType;
+  }
+
+  @Override
+  public boolean apply(@Nullable T input)
+  {
+    if (needsCast) {
+      return castApply(input);
+    }
+    try {
+      return delegate.apply(input);
+    }
+    catch (ClassCastException caster) {
```

Comment:
Will this happen "often"? I ask because exceptions are notoriously expensive, and if this is part of a normal path then it is going to really slow things down.

Reply:
i believe it should happen at most once per predicate creation - so once per segment when used with a value matcher for filtered offset or predicate based index for a bitmap offset, and once per filtered aggregate when used with filtered aggs. I agree they are expensive, that is why i did it like this to catch the exception on the first value that has a problem that throws it and then set the `needsCast` flag to always take the casting path for all checks after that, which is also expensive but I think should be less so than throwing exceptions every row.

Reply:
Also, I want to add that I want to rework this in the future so that it isn't necessary. Today the exceptions should only ever happen with mixed type columns, and I have some ideas on how to make them play nicer with stuff to make this not be an issue, but haven't got to it yet in this PR.

Reply:
OK, got it. I missed that the casting was not going to happen for all rows.